content/blog/buildkite-cd.md

```diff
@@ -0,0 +1,136 @@
+
+# Setting up Buildkite agent
+
+We want to create a separate agent from our normal build agents so that a deployment does not have to wait on any long-running builds. Buildkite agents can be configured to listen to specific queues. The default queue is called `default`, jobs added without a specified queue go onto the `default` queue and Buildkite agents without a specified queue pull jobs from this queue. We can create an agent which listens to a different queue by setting the `queue` tag in the [agent configuration](https://buildkite.com/docs/agent/v3/configuration). For example set `tags="queue=deploy"` to pull jobs from the `deploy` queue.
+
+To run the script on the remote server the agent will need an ssh key pair. These need to be static and not change if the agent is torn down and brought up again. We have added a persistent ssh key pair into our [vault](https://www.vaultproject.io/) secret store. Then when the agent is brought up we read the secret out of the vault and write it into the ssh dir.
+
+```
+AGENT_SSH=~buildkite-agent/.ssh
+mkdir -p $AGENT_SSH
+vault read -field=public secret/deploy/ssh > $AGENT_SSH/id_deploy.pub
+vault read -field=private secret/deploy/ssh > $AGENT_SSH/id_deploy
+chmod 600 $AGENT_SSH/id_deploy
+chown -R buildkite-agent.buildkite-agent $AGENT_SSH
+```
+
+We then add the public key to the list of authorised keys on the app host server.
+
+We can now start the agent. In the agents list on Buildkite, you should see the agent running with the queue set to `deploy`. We now have an agent listening on a dedicated deploy queue that can run scripts on the app host server. Next, we need to set up the deployment pipelines themselves.
+<img src="/img/buildkite-cd-agent.png" alt="png of agent list"/>
+
+# Deployment pipeline
+
+Create a repository for your deployment pipelines and add a pipeline yml at `./buildkite/deploy-pipeline.yml` with content
+
+```
+steps:
+  - label: ":rocket: deploy"
+    command: >-
+      ssh -i ~/.ssh/id_deploy -oStrictHostKeyChecking=accept-new <username>@<host> './deploy.sh'
+    agents:
+      queue: "deploy"
+```
+
+The important parts here are
+* The `ssh` command sets the key to use via `-i ~/.ssh/id_deploy`
+* `StrictHostKeyChecking` is set to `accept-new`, this will accept the host key the first time but refuse to connect if the saved key does not match. This will suffice for us because our agents and host server are on a private internal network so we can trust accepting an unknown key the first time we login to the remote server.
+* `<username>@<host>` should be the username and host of your remote server where the app will be deployed
+* `./deploy.sh` is the name of the deployment script on your remote server
+
+Now we need to add the pipeline to Buildkite. This requires a couple of steps different than adding a new CI build because we want finer control over when this pipeline is triggered.
```

Comment:
```suggestion
Now we need to add the pipeline to Buildkite. This requires a couple of different steps to adding a new CI build because we want finer control over when this pipeline is triggered.
```

Reply:
Not sure if my suggestion is much better tbh

Reply:
How about "This is a little different to adding a new CI build because.."